pkg/mcs/resource_manager/tenant_client/client.go

```diff
@@ -0,0 +1,863 @@
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+type TenantSideKVInterceptor interface {
+	OnRequestWait(ctx context.Context, resourceGroupName string, info RequestInfo) error
+	OnResponse(ctx context.Context, resourceGroupName string, req RequestInfo, resp ResponseInfo) error
+}
+
+type ResourceGroupProvider interface {
+	ListResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup, error)
+	GetResourceGroup(ctx context.Context, resourceGroupName string) (*rmpb.ResourceGroup, error)
+	AddResourceGroup(ctx context.Context, resourceGroupName string, settings *rmpb.GroupSettings) (string, error)
+	ModifyResourceGroup(ctx context.Context, resourceGroupName string, settings *rmpb.GroupSettings) (string, error)
+	DeleteResourceGroup(ctx context.Context, resourceGroupName string) (string, error)
+	AcquireTokenBuckets(ctx context.Context, request *rmpb.TokenBucketsRequest) ([]*rmpb.TokenBucketResponse, error)
+}
+
+func NewResourceGroupController(
+	provider ResourceGroupProvider,
+) (*resourceGroupsController, error) {
+	return newTenantSideCostController(provider)
+}
+
+var _ TenantSideKVInterceptor = (*resourceGroupsController)(nil)
+
+type resourceGroupsController struct {
+	instanceFingerprint string
+	provider            ResourceGroupProvider
+	groupsController    sync.Map
+	config              *Config
+
+	// responseChan is used to receive results from token bucket requests, which
+	// are run in a separate goroutine. A nil response indicates an error.
+	responseChan chan []*rmpb.TokenBucketResponse
+
+	// lowRUNotifyChan is used when the number of available resource is running low and
+	// we need to send an early token bucket request.
+	lowRUNotifyChan chan struct{}
+
+	run struct {
+		now             time.Time
+		lastRequestTime time.Time
+		// requestInProgress is true if we are in the process of sending a request;
+		// it gets set to false when we process the response (in the main loop),
+		// even in error cases.
+		requestInProgress bool
+
+		// requestNeedsRetry is set if the last token bucket request encountered an
+		// error. This triggers a retry attempt on the next tick.
+		//
+		// Note: requestNeedsRetry and requestInProgress are never true at the same
+		// time.
+		requestNeedsRetry bool
+
+		// targetPeriod stores the value of the TargetPeriodSetting setting at the
+		// last update.
+		targetPeriod time.Duration
+	}
+}
+
+func newTenantSideCostController(provider ResourceGroupProvider) (*resourceGroupsController, error) {
+	return &resourceGroupsController{
+		provider:        provider,
+		config:          DefaultConfig(),
+		lowRUNotifyChan: make(chan struct{}, 1),
+		responseChan:    make(chan []*rmpb.TokenBucketResponse, 1),
+	}, nil
+}
+
+func (c *resourceGroupsController) Start(ctx context.Context, instanceFingerprint string) error {
+	if len(instanceFingerprint) == 0 {
+		return errors.New("invalid SQLInstanceID")
+	}
+	c.instanceFingerprint = instanceFingerprint
+	// just for demo
```

Comment:
Should we stll keep this?

Reply:
I think should remove it. will add default group later, current should create new group and then alter user to this group.